Add Linkchecker github action

BlackstoneEngineering opened the pull request:
This PR adds a github action that will search the boxes.treasuredata.com site for broken links and report them if found via a slack integration to the internal TD team

BlackstoneEngineering commented:
@takuti - please review and merge if no issues found. I cannot run the action without merging it into master.

chezou commented:
If you want to debug  this workflow, you can just write temporal condition like `on: [push]` and check the behavioir.
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-a-single-event

BlackstoneEngineering commented:
@chezou - i've added a manual trigger, this is a known working workflow (see the docs repo here for where I pulled it from: https://github.com/treasure-data/TD-API-Documentation/blob/internal/.github/workflows/linkchecker.yml )

This workflow has 2 triggers, one for running monthly on a cron job, and another for manual testing. If its not on the master branch i cannot trigger the manual test. I could switch this to a 'on push' condition to get it to run from a branch, but i'd rather just get it merged into master.

chezou commented:
Ah, didn't know the manual trigger https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/

Don't you need to debug before merging, just in case? I know commit logs may become dirty, but I feel safer since reviewers also can check the GH action's result.

BlackstoneEngineering commented:
It has already been debugged previously. You can see it here: https://github.com/treasure-data/TD-API-Documentation/commits/internal/.github/workflows/linkchecker.yml . I removed it in the most recent commit, but it was in that repo until i moved it here. 

I would be happy to run it, but i cannot run it manually from a branch, it has to be merged into master to be runnable via a manual trigger.

chezou commented on .github/workflows/LinkChecker.yml:
```diff
@@ -0,0 +1,80 @@
+          name: boxes.treasuredata.com.html
+          path: ./logs/boxes.treasuredata.com.html
+
+#       - name: Send email
+#         uses: dawidd6/action-send-mail@v2
+#         with:
+#           server_address: smtp.gmail.com
+#           server_port: 465
+#           username: ${{secrets.MAIL_USERNAME}}
+#           password: ${{secrets.MAIL_PASSWORD}}
+#           subject: Linchecker results for boxes.treasuredata.com
+#           # Literal body:
+#           # body: LinkChecker for boxes.treasuredata.com
+#           # Read file contents as body:
+#           body: file://./logs/boxes.treasuredata.com.html
+#           to: kitazawa@treasure-data.com
+#           from: ci-bot@treasure-data.com # <user@example.com>
+#           # Optional carbon copy recipients
+#           # cc: austin.blackstone@treasure-data.com
+#           # Optional blind carbon copy recipients
+#           # bcc: project-lead@treasure-data.com
+#           # Optional unsigned/invalid certificates allowance:
+#           # ignore_cert: false
+#           # Optional content type (defaults to text/plain):
+#           content_type: text/html
+#           # Optional converting Markdown to HTML (set content_type to text/html too):
+#           convert_markdown: true
+#           # Optional attachments:
+#           attachments: ./logs/boxes.treasuredata.com.html
```
Can you remove unused lines?

```suggestion
```

chezou replied:
Understood. However, we can revert the commit of removal like `git revert COMMIT_HASH` when we need to reuse it. I don't think keeping commented out is mandatory.

takuti replied:
@BlackstoneEngineering I echo what @chezou said; let's remove the comments & keep the workflow clean. Since this is a public repository anyone can refer and contribute, I strongly recommend NOT to have the unused lines. In particular, things like `{kitazawa, ci-bot, austin.blackstone, project-lead}@treasure-data.com` are too much irrelevant information for now.

If you want to have a future reference, simply suppress the comments to a single line like:

> \# Use https://github.com/dawidd6/action-send-mail to alternate Slack notification with email

Don't worry; when you delete the comments with a new commit e.g., `Remove unused email workflow`, the commit hash and corresponding diff tells future contributors how they can send email in the workflow.

Everything else looks good to me.

chezou reviewed (changes requested):
I intended to run for checking treasure-boxes, not TD-API-Documentation, before merging. Anyway, remove unused lines before merging it.

BlackstoneEngineering commented:
Technically yes, practically no, once removed it will be forgotten. I
recommend leaving it in case the feature is needed in the future so it can
be easily enabled. This is future proofing, it has a purpose. It doesn't
matter if it stays, so why not leave it?

On Wed, Apr 21, 2021, 3:32 AM Aki Ariga ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In .github/workflows/LinkChecker.yml
> <https://github.com/treasure-data/treasure-boxes/pull/304#discussion_r616940967>
> :
>
> > +#       - name: Send email
> +#         uses: ***@***.***
> +#         with:
> +#           server_address: smtp.gmail.com
> +#           server_port: 465
> +#           username: ${{secrets.MAIL_USERNAME}}
> +#           password: ${{secrets.MAIL_PASSWORD}}
> +#           subject: Linchecker results for boxes.treasuredata.com
> +#           # Literal body:
> +#           # body: LinkChecker for boxes.treasuredata.com
> +#           # Read file contents as body:
> +#           body: file://./logs/boxes.treasuredata.com.html
> +#           to: ***@***.***
> +#           from: ***@***.*** # ***@***.***>
> +#           # Optional carbon copy recipients
> +#           # cc: ***@***.***
> +#           # Optional blind carbon copy recipients
> +#           # bcc: ***@***.***
> +#           # Optional unsigned/invalid certificates allowance:
> +#           # ignore_cert: false
> +#           # Optional content type (defaults to text/plain):
> +#           content_type: text/html
> +#           # Optional converting Markdown to HTML (set content_type to text/html too):
> +#           convert_markdown: true
> +#           # Optional attachments:
> +#           attachments: ./logs/boxes.treasuredata.com.html
>
> Understood. However, we can revert the commit of removal like git revert
> COMMIT_HASH when we need to reuse it. I don't think keeping commented out
> is mandatory.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/treasure-data/treasure-boxes/pull/304#discussion_r616940967>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABG3PU7QMA43KKUHGR6YP7LTJXCD5ANCNFSM43HQ2FNA>
> .
>

BlackstoneEngineering commented:
Updated, please merge at your convenience.
